pkg/agent/controller/networkpolicy/node_reconciler_linux.go

```diff
@@ -0,0 +1,703 @@
+	ingressCoreIPTRules := make(map[iptables.Protocol][]*coreIPTRule)
+	egressCoreIPTRules := make(map[iptables.Protocol][]*coreIPTRule)
+
+	for _, rule := range rules {
+		iptRules, lastRealized := r.computeIPTRules(rule)
+		ruleID := rule.ID
+		for ipProtocol, iptRule := range iptRules {
+			// Sync all ipsets.
+			if iptRule.IPSet != "" {
+				if err := r.routeClient.AddOrUpdateNodeNetworkPolicyIPSet(iptRule.IPSet, iptRule.IPSetMembers, iptRule.IsIPv6); err != nil {
+					return err
+				}
+			}
+			// Collect all service iptables rules and chains.
+			if iptRule.ServiceIPTChain != "" {
+				serviceIPTChains[ipProtocol] = append(serviceIPTChains[ipProtocol], iptRule.ServiceIPTChain)
+				serviceIPTRules[ipProtocol] = append(serviceIPTRules[ipProtocol], iptRule.ServiceIPTRules)
+			}
+
+			// Collect all core iptables rules.
+			coreIPTRule := &coreIPTRule{ruleID, iptRule.Priority, iptRule.CoreIPTRule}
+			if rule.Direction == v1beta2.DirectionIn {
+				ingressCoreIPTRules[ipProtocol] = append(ingressCoreIPTRules[ipProtocol], coreIPTRule)
+			} else {
+				egressCoreIPTRules[ipProtocol] = append(egressCoreIPTRules[ipProtocol], coreIPTRule)
+			}
+		}
+		lastRealizeds[ruleID] = lastRealized
+	}
+	for _, ipProtocol := range r.ipProtocols {
+		isIPv6 := iptables.IsIPv6Protocol(ipProtocol)
+		if err := r.routeClient.AddOrUpdateNodeNetworkPolicyIPTables(serviceIPTChains[ipProtocol], serviceIPTRules[ipProtocol], isIPv6); err != nil {
+			return err
+		}
+		if err := r.addOrUpdateCoreIPTRules(ingressCoreIPTRules[ipProtocol], config.NodeNetworkPolicyIngressRulesChain, isIPv6, false); err != nil {
+			return err
+		}
+		if err := r.addOrUpdateCoreIPTRules(egressCoreIPTRules[ipProtocol], config.NodeNetworkPolicyEgressRulesChain, isIPv6, false); err != nil {
+			return err
+		}
+	}
+
+	for ruleID, lastRealized := range lastRealizeds {
+		r.lastRealizeds.Store(ruleID, lastRealized)
+	}
+	return nil
+}
+
+func (r *nodeReconciler) Forget(ruleID string) error {
+	klog.InfoS("Forgetting rule", "rule", ruleID)
+
+	value, exists := r.lastRealizeds.Load(ruleID)
+	if !exists {
+		return nil
+	}
+
+	lastRealized := value.(*nodePolicyLastRealized)
+	coreIPTChain := lastRealized.coreIPTChain
+
+	for _, ipProtocol := range r.ipProtocols {
+		isIPv6 := iptables.IsIPv6Protocol(ipProtocol)
+		if err := r.deleteCoreIPRule(ruleID, coreIPTChain, isIPv6); err != nil {
+			return err
+		}
+		if lastRealized.ipsets[ipProtocol] != "" {
+			if err := r.routeClient.DeleteNodeNetworkPolicyIPSet(lastRealized.ipsets[ipProtocol], isIPv6); err != nil {
+				return err
+			}
+		}
+		if lastRealized.serviceIPTChain != "" {
+			if err := r.routeClient.DeleteNodeNetworkPolicyIPTables([]string{lastRealized.serviceIPTChain}, isIPv6); err != nil {
+				return err
+			}
+		}
+	}
+
+	r.lastRealizeds.Delete(ruleID)
+	return nil
+}
+
+func (r *nodeReconciler) GetRuleByFlowID(ruleFlowID uint32) (*types.PolicyRule, bool, error) {
+	return nil, false, nil
```

Comment:
I don't get the purpose of this function. Why does it always return false here? Is this for testing only?

Reply:
We need to implement this interface
```go
// Reconciler is an interface that knows how to reconcile the desired state of
// CompletedRule with the actual state of Openflow entries.
type Reconciler interface {
	// Reconcile reconciles the desired state of the provided CompletedRule
	// with the actual state of Openflow entries.
	Reconcile(rule *CompletedRule) error

	// BatchReconcile reconciles the desired state of the provided CompletedRules
	// with the actual state of Openflow entries in batch. It should only be invoked
	// if all rules are newly added without last realized status.
	BatchReconcile(rules []*CompletedRule) error

	// Forget cleanups the actual state of Openflow entries of the specified ruleID.
	Forget(ruleID string) error

	// GetRuleByFlowID returns the rule from the async rule cache in idAllocator cache.
	GetRuleByFlowID(ruleID uint32) (*types.PolicyRule, bool, error)

	// RunIDAllocatorWorker runs the worker that deletes the rules from the cache
	// in idAllocator.
	RunIDAllocatorWorker(stopCh <-chan struct{})
}

```